// ionic4-auto-complete/esm2015/main.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { FormsModule } from '@angular/forms';
import { IonicModule } from '@ionic/angular';
import { AutoCompleteComponent } from './auto-complete/auto-complete.component';
import { BoldPrefix } from './bold-prefix.pipe';
export * from './auto-complete-options.model';
export * from './auto-complete-styles.model';
export * from './auto-complete.service';
export * from './auto-complete/auto-complete.component';
export * from './bold-prefix.pipe';
export class AutoCompleteModule {
    static forRoot() {
        return {
            ngModule: AutoCompleteModule,
            providers: []
        };
    }
}
AutoCompleteModule.decorators = [
    { type: NgModule, args: [{
                declarations: [
                    AutoCompleteComponent,
                    BoldPrefix
                ],
                exports: [
                    AutoCompleteComponent,
                    BoldPrefix
                ],
                imports: [
                    CommonModule,
                    FormsModule,
                    IonicModule
                ]
            },] }
];
